### ISSUE TITLE
javac front-end exception thrown when building guava-14.0.1

### ISSUE THREAD

**Issue by guilhermesgb** [google/j2objc#838]
Hello guys, thanks again for the help with issue #836! Your support here is really top-notch.

I've stumbled upon a pretty interesting finding.

A bit of story for context first, though:

After getting past that problem by building from source [the version of j2obc at this point in history](https://github.com/google/j2objc/commit/564997a30d5b8f6e96fc248a97279abad5a77760) and using it to translate my project, I realized that, unfortunately, I still got problems with the translation process.

This is the issue I got, pretty nasty IMHO as it was pretty hard to track down what caused it:

```
$ ./j2objc_translate
Obtaining Java sources list...
Preparing to translate to iOS-ready Objective-C sources...
Translating using classpath: lib/guava-14.0.1-sources.jar:lib/slf4j-api-1.7.13-sources.jar:src.

Exception in thread "main" com.sun.tools.javac.util.Abort
	at com.sun.tools.javac.comp.Check.completionError(Check.java:287)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:584)
	at com.sun.tools.javac.comp.Attr.attribType(Attr.java:638)
	at com.sun.tools.javac.comp.Attr.attribType(Attr.java:631)
	at com.sun.tools.javac.comp.Attr.visitTypeApply(Attr.java:3911)
	at com.sun.tools.javac.tree.JCTree$JCTypeApply.accept(JCTree.java:2135)
	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
	at com.sun.tools.javac.comp.Attr.attribType(Attr.java:638)
	at com.sun.tools.javac.comp.Attr.attribType(Attr.java:631)
	at com.sun.tools.javac.comp.Attr.attribBase(Attr.java:786)
	at com.sun.tools.javac.comp.MemberEnter.complete(MemberEnter.java:1057)
	at com.sun.tools.javac.code.Symbol.complete(Symbol.java:574)
	at com.sun.tools.javac.code.Symbol$ClassSymbol.complete(Symbol.java:1037)
	at com.sun.tools.javac.comp.Enter.complete(Enter.java:493)
	at com.sun.tools.javac.comp.Enter.main(Enter.java:471)
	at com.sun.tools.javac.main.JavaCompiler.enterTrees(JavaCompiler.java:982)
	at com.sun.tools.javac.api.JavacTaskImpl.enter(JavacTaskImpl.java:316)
	at com.sun.tools.javac.api.JavacTaskImpl.analyze(JavacTaskImpl.java:359)
	at com.sun.tools.javac.api.JavacTaskImpl.analyze(JavacTaskImpl.java:344)
	at com.google.devtools.j2objc.javac.JavacParser.parseFiles(JavacParser.java:160)
	at com.google.devtools.j2objc.pipeline.FileProcessor.processBatch(FileProcessor.java:141)
	at com.google.devtools.j2objc.pipeline.FileProcessor.processInputs(FileProcessor.java:70)
	at com.google.devtools.j2objc.pipeline.TranslationProcessor.processInputs(TranslationProcessor.java:77)
	at com.google.devtools.j2objc.J2ObjC.run(J2ObjC.java:128)
	at com.google.devtools.j2objc.J2ObjC.main(J2ObjC.java:178)
 ```

The stacktrace gave no clue about which portion of my project that exactly caused it, as opposed to issue #836's situation.

Googling wouldn't help me much. The closest I got to finding someone facing the exact same problem was [this thread](https://github.com/clojure-emacs/cider-nrepl/issues/130). I don't even know if they really had the same problem, I didn't  bother looking to much into their case as it seems to be a completely unrelated project.

Eventually I could narrow down the problem a bit more... noticed that it would happen when I tried to include **guava-sources.jar version 19.0** into my `-sourcepath`. So I thought that maybe it was a problem with the version I was using, and I got evidence to support my hyphothesis when I found [this j2objc-related project including a subtly customized version of guava](https://github.com/jefflovejapan/j2objc/blob/master/guava/guava-14.0.1-sources.jar).

But using it still wouldn't help me. The problem apparently still happened while j2objc attempted at translating the guava sources (or something along those lines).

So my interesting finding is this: when I decided to give it a try at using the distributed **j2objc-1.3.1** version to translate the exact same project, it actually worked!! So probably some commit after the distributed 1.3.1 version is causing this issue.

You guys can use [this repository](https://github.com/guilhermesgb/j2objc-sample-project) for verifying the issue on your own. If you decide to test it using the `j2objc_translate` script I provided you might want to have your `$PATH` point to the *latest version of j2objc on the master branch* and also edit the file to point the proper absolute path of your **j2objc-1.3.1** tool.

Thanks again for all the help!

PS: Something I might try at some point is a `git-bisect` for finding which commit seems to be the guilty party. I would be a fun occasion to actually learn how it works in practice :)

Cheers!

**Comment by kstanger** [comment]
Before we jump into this too deeply, there are a few things you could try:

Since 1.3.1 we have switched our default parsing front end from JDT to Javac ([here](https://github.com/google/j2objc/commit/02db36ef3f23a38e9487705951a06e0ad4311999)). You can toggle between the two using `-Xuse-jdt` and `-Xuse-javac` or by setting an environment variable `J2OBJC_FRONT_END=JDT` or `=JAVAC`. Switching back to JDT will probably unblock you for now, but we plan to get rid of JDT eventually so we should make sure Javac works for you.

Since the error is coming from Javac, it's very likely that you would get the same error when using javac to compile the same code. Try switching "j2objc" with "javac" in your command line and see if you get the same error. (When building locally, j2objc will be using the same version of javac that is on your machine, taken from $JAVA_HOME/lib/tools.jar)

I believe that Guava 19 is the last version written in Java7. Try toggling between `-source 7` and `-source 8` and see if either of those help.

**Comment by tomball** [comment]
guava-*14.0.1*-sources.jar? Please consider upgrading, as this release is
almost 4 years old. As Keith said, it's likely that the version of javac on
your system cannot compile it (though do try the -source 1.7 suggestion
first).

**Comment by guilhermesgb** [comment]
Thank you all for the quick responses. So, I did as you suggested, got back to using Guava 19, and attempted at translating with the `-source` flags set first -- it yields the same results.

(Weirdly enough, I didn't get quite the exact same errors using `javac` directly, but it didn't work nonetheless.)

But Keith was spot on in his suggestion of swapping the front end. Using JDT really did work as expected! So I'll go ahead and experiment trying out the next steps. Please et me know if you need anything from me regarding this issue.

**Comment by kstanger** [comment]
You might consider updating your java installation if it is older than Guava 19. I'd be interested to see the errors you get from javac.

**Comment by guilhermesgb** [comment]
I do have a up-to-date javac version, though.. let me know if I'm doing something wrong. When I try running it:

```
#!/bin/zsh

CLASSPATH="$(for i in lib/*.jar ; do echo -n $i: ; done)src"

echo "Obtaining Java sources list..."
find . \! -name ".java" | grep \\.java > java_sources.list
echo "Preparing to translate to iOS-ready Objective-C sources..."
mkdir -p j2objc_gen/src
echo "Translating using classpath: `echo $CLASSPATH`."
javac -d j2objc_gen/src -sourcepath "$CLASSPATH" @java_sources.list
echo "Done translating."
rm java_sources.list
```

The following errors pop up everywhere (I tried to filter repeated occurrences of the same kinds of errors):

```
Obtaining Java sources list...
Preparing to translate to iOS-ready Objective-C sources...
Translating using classpath: lib/guava-19.0-sources.jar:lib/jsr305-1.3.9.jar:lib/slf4j-api-1.7.13-sources.jar:src.
lib/guava-19.0-sources.jar(com/google/common/eventbus/SubscriberRegistry.java):38: error: package com.google.j2objc.annotations does not exist
import com.google.j2objc.annotations.Weak;
                                    ^
lib/guava-19.0-sources.jar(com/google/common/eventbus/Subscriber.java):22: error: package com.google.j2objc.annotations does not exist
import com.google.j2objc.annotations.Weak;
                                    ^
```
[...]
```
lib/guava-19.0-sources.jar(com/google/common/collect/AbstractMapBasedMultimap.java):354: error: cannot find symbol
  @WeakOuter
   ^
  symbol:   class WeakOuter
  location: class AbstractMapBasedMultimap<K,V>
  where K,V are type-variables:
    K extends Object declared in class AbstractMapBasedMultimap
    V extends Object declared in class AbstractMapBasedMultimap
lib/guava-19.0-sources.jar(com/google/common/collect/Maps.java):2800: error: cannot find symbol
    @WeakOuter
     ^
  symbol:   class WeakOuter
  location: class FilteredEntryMap<K,V>
  where K,V are type-variables:
    K extends Object declared in class FilteredEntryMap
    V extends Object declared in class FilteredEntryMap
lib/guava-19.0-sources.jar(com/google/common/collect/Maps.java):2834: error: cannot find symbol
    @WeakOuter
     ^
  symbol:   class WeakOuter
  location: class FilteredEntryMap<K,V>
  where K,V are type-variables:
    K extends Object declared in class FilteredEntryMap
    V extends Object declared in class FilteredEntryMap
lib/guava-19.0-sources.jar(com/google/common/collect/Maps.java):2910: error: cannot find symbol
    @WeakOuter
     ^
  symbol:   class WeakOuter
  location: class FilteredEntrySortedMap<K,V>
  where K,V are type-variables:
    K extends Object declared in class FilteredEntrySortedMap
    V extends Object declared in class FilteredEntrySortedMap
lib/guava-19.0-sources.jar(com/google/common/collect/ImmutableMap.java):536: error: cannot find symbol
  @WeakOuter
   ^
```
[...]
```
  symbol:   class WeakOuter
  location: class ImmutableMultiset<E>
  where E is a type-variable:
    E extends Object declared in class ImmutableMultiset
lib/guava-19.0-sources.jar(com/google/common/cache/LocalCache.java):4495: error: cannot find symbol
  @WeakOuter
   ^
```
[...]
```
  symbol:   class WeakOuter
  location: class LocalCache<K,V>
  where K,V are type-variables:
    K extends Object declared in class LocalCache
    V extends Object declared in class LocalCache
lib/guava-19.0-sources.jar(com/google/common/util/concurrent/CollectionFuture.java):26: error: package com.google.j2objc.annotations does not exist
import com.google.j2objc.annotations.WeakOuter;
                                    ^
lib/guava-19.0-sources.jar(com/google/common/util/concurrent/AggregateFuture.java):27: error: package com.google.j2objc.annotations does not exist
import com.google.j2objc.annotations.WeakOuter;
                                    ^
lib/guava-19.0-sources.jar(com/google/common/util/concurrent/AggregateFuture.java):90: error: cannot find symbol
  @WeakOuter
   ^
  symbol:   class WeakOuter
  location: class AggregateFuture<InputT,OutputT>
  where InputT,OutputT are type-variables:
    InputT extends Object declared in class AggregateFuture
    OutputT extends Object declared in class AggregateFuture
lib/guava-19.0-sources.jar(com/google/common/util/concurrent/CollectionFuture.java):38: error: cannot find symbol
  @WeakOuter
   ^
  symbol:   class WeakOuter
  location: class CollectionFuture<V,C>
  where V,C are type-variables:
    V extends Object declared in class CollectionFuture
    C extends Object declared in class CollectionFuture
lib/guava-19.0-sources.jar(com/google/common/collect/AbstractMapBasedMultimap.java):614: error: cannot find symbol
  @WeakOuter
   ^
```
[...]
```
  symbol:   class WeakOuter
  location: class AbstractMapBasedMultimap<K,V>
  where K,V are type-variables:
    K extends Object declared in class AbstractMapBasedMultimap
    V extends Object declared in class AbstractMapBasedMultimap
lib/guava-19.0-sources.jar(com/google/common/base/AbstractIterator.java):31: error: duplicate class: com.google.common.base.AbstractIterator
abstract class AbstractIterator<T> implements Iterator<T> {
         ^
lib/guava-19.0-sources.jar(com/google/common/collect/Collections2.java):490: error: cannot access AbstractIterator
  private static final class OrderedPermutationIterator<E> extends AbstractIterator<List<E>> {
                                                                   ^
  bad source file: lib/guava-19.0-sources.jar(com/google/common/collect/AbstractIterator.java)
    file does not contain class com.google.common.collect.AbstractIterator
    Please remove or make sure it appears in the correct subdirectory of the sourcepath.
lib/guava-19.0-sources.jar(com/google/common/eventbus/SubscriberRegistry.java):71: error: cannot find symbol
  @Weak private final EventBus bus;
   ^
  symbol:   class Weak
  location: class SubscriberRegistry
lib/guava-19.0-sources.jar(com/google/common/eventbus/Subscriber.java):51: error: cannot find symbol
  @Weak private EventBus bus;
   ^
  symbol:   class Weak
  location: class Subscriber
lib/guava-19.0-sources.jar(com/google/common/cache/LocalCache.java):2014: error: cannot find symbol
    @Weak final LocalCache<K, V> map;
     ^
  symbol:   class Weak
  location: class Segment<K,V>
  where K,V are type-variables:
    K extends Object declared in class Segment
    V extends Object declared in class Segment
lib/guava-19.0-sources.jar(com/google/common/collect/Maps.java):3794: error: cannot find symbol
    @Weak final Map<K, V> map;
     ^
  symbol:   class Weak
  location: class Values<K,V>
  where K,V are type-variables:
    K extends Object declared in class Values
    V extends Object declared in class Values
lib/guava-19.0-sources.jar(com/google/common/collect/Maps.java):3614: error: cannot find symbol
    @Weak final Map<K, V> map;
     ^
  symbol:   class Weak
  location: class KeySet<K,V>
  where K,V are type-variables:
    K extends Object declared in class KeySet
    V extends Object declared in class KeySet
lib/guava-19.0-sources.jar(com/google/common/collect/ImmutableSetMultimap.java):433: error: cannot find symbol
    @Weak private final transient ImmutableSetMultimap<K, V> multimap;
     ^
  symbol:   class Weak
  location: class EntrySet<K,V>
  where K,V are type-variables:
    K extends Object declared in class EntrySet
    V extends Object declared in class EntrySet
lib/guava-19.0-sources.jar(com/google/common/collect/ImmutableMultimap.java):517: error: cannot find symbol
    @Weak final ImmutableMultimap<K, V> multimap;
     ^
  symbol:   class Weak
  location: class EntryCollection<K,V>
  where K,V are type-variables:
    K extends Object declared in class EntryCollection
    V extends Object declared in class EntryCollection
lib/guava-19.0-sources.jar(com/google/common/collect/ImmutableMultimap.java):660: error: cannot find symbol
    @Weak private final transient ImmutableMultimap<K, V> multimap;
     ^
  symbol:   class Weak
  location: class Values<K,V>
  where K,V are type-variables:
    K extends Object declared in class Values
    V extends Object declared in class Values
lib/guava-19.0-sources.jar(com/google/common/cache/LocalCache.java):4453: error: cannot find symbol
    @Weak final ConcurrentMap<?, ?> map;
     ^
  symbol:   class Weak
  location: class LocalCache<K,V>.AbstractCacheSet<T>
  where T,K,V are type-variables:
    T extends Object declared in class LocalCache.AbstractCacheSet
    K extends Object declared in class LocalCache
    V extends Object declared in class LocalCache
Done translating.
```

It does feel like I'm missing out on something simple.

**Comment by tomball** [comment]
The current version of Guava [depends on j2objc-annotations](https://github.com/google/guava/blob/master/pom.xml#L212), so add a path to the j2objc_annotations.jar file in your j2objc distribution. The good news is that this dependency means that Guava leaks much less on iOS than in earlier versions.

**Comment by tomball** [comment]
One problem I see with your script is that it appears to confuse -sourcepath with -classpath. The jar files in guava/lib contain class files (*.class) and so should be put on the class path. Only files or directories that contain Java source files belong on the source path. The j2objc_annotations.jar file should be included in the classpath.

I tried compiling guava-14.0.1 with javac and it failed with a Java 8 JVM, whether -source 1.7 or 1.8 is specified, while it does build with a Java 7 JDK. Since j2objc's jre_emul library is based on Java 8 now, that means that version of Guava cannot be translated by j2objc 1.2 (the first release supporting Java 8).

guava-19.0 (and higher) does build with Java 8, and with j2objc. To get the correct jars, I ran "mvn generate-resources" and copied the classpath jars to a local "jars" directory):

```
$ javac -d build \
    -classpath jars/jsr305.jar:jars/javax-inject.jar:jars/j2objc_annotations.jar:jars/animal-sniffer-annotations-1.14.jar \
    -sourcepath guava/src `find guava/src -name '*.java'`
$ j2objc -d build \
    -classpath jars/jsr305.jar:jars/javax-inject.jar:jars/j2objc_annotations.jar:jars/animal-sniffer-annotations-1.14.jar \
    -sourcepath guava/src `find guava/src -name '*.java'`
```
